common/taskqueue/task_queue.go

```diff
@@ -0,0 +1,359 @@
+// The MIT License
+//
+// Copyright (c) 2020 Temporal Technologies Inc.  All rights reserved.
+//
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package taskqueue
+
+import (
+	"errors"
+	"fmt"
+	"strconv"
+	"strings"
+
+	enumspb "go.temporal.io/api/enums/v1"
+	taskqueuepb "go.temporal.io/api/taskqueue/v1"
+	"go.temporal.io/server/common/namespace"
+)
+
+const (
+	// NonRootPartitionPrefix is the prefix for all mangled task queue names.
+	NonRootPartitionPrefix = "/_sys/"
+	PartitionDelimiter     = "/"
+)
+
+type (
+
+	// TaskQueue can be NormalTaskQueue or StickyTaskQueue.
+	TaskQueue interface {
```

Comment:
I'd love if we could avoid repetitive names like `taskqueue.TaskQueue`, what about names like `taskqueue.Interface` with implementations `taskqueue.Normal` and `taskqueue.Sticky`? You can see discussion of this on the [official go blog](https://go.dev/blog/package-names):

> Avoid repetition. Since client code uses the package name as a prefix when referring to the package contents, the names for those contents need not repeat the package name. The HTTP server provided by the http package is called Server, not HTTPServer. Client code refers to this type as http.Server, so there is no ambiguity.

Reply:
Great feedback @tdeebswihart. I'm going to rename the types as you suggested, except I think I like to keep `taskqueue.TaskQueue` as is, renaming to `taskqueue.Interface` feels a little awkward to me. `pkg.Pkg` is mentioned in the link you referred and not discouraged. Many packages use that naming, e.g. `time.Time`.